### ISSUE TITLE
Document performance expectations

### ISSUE THREAD

**Issue by bzz** [src-d/gemini#42]
Document in README the resources, needed to process 1k, 2k, 10k (and 100k speculative) of the .siva files

So good start would be 
 - document the known configuration of the cluster
 - running it, documenting how long it takes to finish, 
 - what CPU-load, Mem, IO-thoughtput workload it creates on that cluster (using Sysdig)
 - see what resource is the bottleneck
 - optimize, in order to utilize that resource better (i.e in case of throughput - have more executor JVMs running on the same machine)
 - extrapolate the expectations on the resources needed if one wants to process 100k (and a timeframe)

**Comment by smacker** [comment]
> optimize, in order to utilize that resource better (i.e in case of throughput - have more executor JVMs running on the same machine)

how do to that? we don't control spark cluster.

**Comment by bzz** [comment]
> how do to that? we don't control spark cluster.

let's measure, identify and document the bottleneck first and then discuss the possible options that we might have. I.e this can be powerful argument for changing https://github.com/src-d/charts/tree/master/spark to [Apache Spark on k8s](https://dist.apache.org/repos/dist/dev/spark/v2.3.0-rc1-docs/_site/running-on-kubernetes.html).

**Comment by smacker** [comment]
Blocked: HDFS is down. No ETA for now.

**Comment by smacker** [comment]
Hit https://github.com/src-d/engine/issues/323

**Comment by bzz** [comment]
Thanks for keeping it updated! 

BTW, super-nice issue description and example how to reproduce.

**Comment by bzz** [comment]
Engine issue is resolved in https://github.com/src-d/engine/releases/tag/v0.5.1

**Comment by smacker** [comment]
yep. But the engine api has changed a bit. We need to update gemini.

**Comment by smacker** [comment]
Run gemini on new 1k dataset with new engine. And it works!!!!

The bad new is timing: 24 min.
I don't really know how to profile it, but I saw that only 1 job is taking much time, most probably there is 1 huge repo.

**Comment by smacker** [comment]
10k has failed with https://github.com/src-d/engine/issues/332

**Comment by smacker** [comment]
currently is blocked by https://github.com/src-d/engine/issues/336

**Comment by bzz** [comment]
To move this forward, can we please submit a PR to engine that loggs `RevWalkException`, same way as [MissingObjectException](https://github.com/src-d/engine/blob/master/src/main/scala/tech/sourced/engine/iterator/CommitIterator.scala?utf8=%E2%9C%93#L146) is handled and test with a custom build of engine from this PR to avoid waiting for a release?

**Comment by bzz** [comment]
@carlosms could you please check if https://github.com/src-d/engine/pull/347 solves the issue and allows us to move forward with https://github.com/src-d/gemini/issues/42 ?

If that PR is tested on real data and solves the issue -  it may be worth posting this information on the PR as well.

**Comment by bzz** [comment]
Title and description are updated to represent the current goal.

**Comment by smacker** [comment]
10k repos are processed successfully with engine 0.5.7. Full PGA is failing with OOM with default params. Need to tune them.

**Comment by bzz** [comment]
Plan is

 - make sure PGA is available on staging cluster
 - run Gemini \w latest Engine 0.6.1 on 1k, 10k, 100k, 200k

**Comment by bzz** [comment]
PGA download is finished 🎉 but it's a bit :suspect: as only 2.4Tb not 2.7Tb as the rumor has it to be. Would verify PGA integrity first with https://github.com/src-d/datasets/issues/53

**Comment by bzz** [comment]
Pre-conditions for running new Gemini on pipeline staging Apache Spark cluster:
 - [ ] PGA downloaded to HDFS (WIP)
 - [ ] to deploy Feature Extractors, same as bblfsh, collocated with Spark Workers
        we need to provide right CLI arguments for [spark cluster helm](https://github.com/src-d/infrastructure/blob/master/docs/pipeline-staging/cluster_applications_installation_notes.md#spark-cluster)
        @rporres suggestion is to test [yml](https://github.com/src-d/infrastructure/blob/master/docs/pipeline-staging/cluster_applications_installation_notes.md#spark-cluster) locally with Minikube first
        New issue for https://github.com/src-d/issues-infrastructure/issues would be needed, to apply this configuration to the full Apache Spark cluster

**Comment by bzz** [comment]
blocked by src-d/backlog#1266

**Comment by bzz** [comment]
- FS are [running](http://127.0.0.1:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/#/pod/feature-extractor/fe-spark-spark-worker-748277027-1fsvk?namespace=feature-extractor)
 - `pga get` second round, ETA `8h52m56`

**Comment by bzz** [comment]
Full PGA was downloaded to HDFS 🎉  https://github.com/src-d/datasets/issues/53#issuecomment-396528917
```
$ zgrep -o "[0-9a-z]*\.siva" ~/.pga/latest.csv.gz | sort | uniq | wc -l
239807

$ hdfs dfs -ls -R hdfs://hdfs-namenode/pga/siva/latest | grep -c "\.siva$"
239807
```

**Comment by bzz** [comment]
Plan

 - [ ] run latest Gemini \w latest Engine 0.6.3 on single shard (~1000 repos, ~1/250 of whole)
 - [ ] run on whole PGA

**Comment by bzz** [comment]
Blocked, as all Feature Extractors are deployed under https://github.com/src-d/issues-infrastructure/issues/184 are part of new, separate Apache Spark cluster in a different k8s namespace `-n feature-extractor`, that does not seem to have access to HDFS 😕

**Comment by bzz** [comment]
1.2h for hashing a ~1/250 of PGA on 3 machines of pipeline staging cluster

Configuration:
<details>
```
  --conf "spark.executor.memory=16g" \
  --conf "spark.local.dir=/spark-temp-data" \
  --conf "spark.executor.extraJavaOptions='-Djava.io.tmpdir=/spark-temp-data -Dlog4j.configuration=log4j.properties'" \
  --conf "spark.driver.memory=8g" \
  --conf "spark.tech.sourced.engine.skip.read.errors=true" \
  --conf "spark.files.maxPartitionBytes=12582912" \
  --conf "spark.serializer=org.apache.spark.serializer.KryoSerializer" \
  --conf "spark.eventLog.enabled=true" \
  --conf "spark.eventLog.dir=hdfs://hdfs-namenode.default.svc.cluster.local/apps/gemini/spark-logs" \
  --files src/main/resources/log4j.properties \
```
</details>

Command:
```
time MASTER="spark://fe-spark-spark-master:7077" ./hash -v \
  -k dockergemini3 \
  -h scylladb.default.svc.cluster.local \
  hdfs://hdfs-namenode.default.svc.cluster.local/pga/siva/latest/ff 2>&1 | tee hash-pga-ff-3.logs
```

Output:
```
```

DB
```
$  kubectl exec -it scylladb-0 -- /bin/bash
$ cqlsh

use dockergemini3;
select count(1) from meta;

select count(1) from hashtables;

```

**Comment by bzz** [comment]
> Question: how are we sampling the repos for each of these tests?

Good question. We always just used only a single shard of PGA dataset - all the repos, who's siva file names start with prefix `/ff/`.

Overall, on Apache Spark performance depends on data distribution A LOT, so attaching .siva file size distribution histogram in 10mb buckets

```
hdfs dfs -du hdfs://hdfs-namenode/pga/siva/latest/ff/ | grep "\.siva$" | awk -v "size=100048576" -f hist.awk
0 100048576 912
100048576 200097152 17
200097152 300145728 2
300145728 400194304 1
400194304 500242880 1
500242880 600291456
600291456 700340032
700340032 800388608
800388608 900437184
900437184 1000485760 1
1000485760 1100534336 1
```

**Comment by bzz** [comment]
Local: 1mb, 30k features
Cluster: 170Mb, 5.5mil features

### DataFrame
local: 8sec, cluster: 4sec
```scala
val freqDf = features.withColumnRenamed("_1", "feature").withColumnRenamed("_2", "doc")
  .select("feature", "doc")
  .distinct
  .groupBy("feature")
  .agg(count("*").alias("cnt"))
  .map(row => (row.getAs[String]("feature"), row.getAs[Long]("cnt")))
  .collect().toMap
```

### RDD
local: 4sec, cluster: 5s
```scala
val freq = features.rdd
  .map { case (feature, doc, _) => (feature, doc) }
  .distinct
  .map { case (token, _) => (token, 1) }
  .reduceByKey(_ + _)
  .collectAsMap()
```

Changing Gemini to use DataFrame API everywhere seems not beneficial.

**Comment by bzz** [comment]
There are 141 .siva files bigger then 1Gb, with rest 260+k being smaller.
Those outliers can be moved, to get shorter tail of task execution time on average

<details>

```
hdfs dfs -ls -R hdfs://hdfs-namenode/pga/siva/latest/ | grep "\.siva$" | awk '{if ($5 > 1073741824) print $8}' | wc -l
141

hdfs dfs -ls -R hdfs://hdfs-namenode/pga/siva/latest/ \
  | grep "\.siva$" | awk '{if ($5 > 1073741824) print $8}' \
  | cut -d '/' -f 7- \
  | xargs -I{} sh -c 'hdfs dfs -mkdir -p $(dirname hdfs://hdfs-namenode/pga/siva/biggest/{}); hdfs dfs -mv hdfs://hdfs-namenode/pga/siva/latest/{} hdfs://hdfs-namenode/pga/siva/biggest/{}'
```
<details>

After removing those, Spark jobs start to fail quickly with 
```
org.apache.spark.SparkException: Kryo serialization failed: Buffer overflow. Available: 0, required: 32343809.
```

After setting `spark.kryoserializer.buffer.max=1g` jobs fail with

```
tech.sourced.siva.SivaException: Exception at file 022c7272f0c1333a536cb319beadc4171cc8ff6a.siva: At Index footer, index size: Java implementation of siva doesn't support values greater than 9223372036854775807
```

which at this point might indicate corrupted .siva files on `pga get`

**Comment by bzz** [comment]
After moving biggest files, jobs fail with

```
org.apache.spark.SparkException: Kryo serialization failed: Buffer overflow. Available: 0, required: 32343809.
```

After setting `spark.kryoserializer.buffer.max=1g` jobs fail with

```
tech.sourced.siva.SivaException: Exception at file 022c7272f0c1333a536cb319beadc4171cc8ff6a.siva: At Index footer, index size: Java implementation of siva doesn't support values greater than 9223372036854775807
```

which at this point might indicate broken .siva files on `pga get`

**Comment by smacker** [comment]
@bzz here is your issue: https://github.com/src-d/engine/issues/414
different file but the same error.

**Comment by bzz** [comment]
Simple processing of full PGA from `/pga2` \w Engine finished in `59.1 h`, using half the cores 🎉 

<details>

```
export SPARK_HOME="/opt/spark-2.2.0-bin-hadoop2.7"
$SPARK_HOME/bin/spark-shell --master "spark://fe-spark-spark-master:7077" \
  --name "Spark shell \w Gemini - PGA2, count file sizes" \
  --conf "spark.executor.memory=8g" \
  --conf "spark.cores.max=144" \
  --conf "spark.hadoop.javax.jdo.option.ConnectionURL=jdbc:derby:;databaseName=metastore_db2;create=true" \
  --conf "spark.local.dir=/spark-temp-data" \
  --conf "spark.executor.extraJavaOptions=-Djava.io.tmpdir=/spark-temp-data -Dlog4j.configuration=log4j.properties -XX:+HeapDumpOnOutOfMemoryError -XX:MaxDirectMemorySize=20G" \
  --conf "spark.driver.memory=8g" \
  --conf "spark.tech.sourced.engine.skip.read.errors=true" \
  --conf "spark.files.maxPartitionBytes=12582912" \
  --conf "spark.serializer=org.apache.spark.serializer.KryoSerializer" \
  --conf "spark.kryoserializer.buffer.max=1g" \
  --conf "spark.eventLog.enabled=true" \
  --conf "spark.eventLog.dir=hdfs://hdfs-namenode.default.svc.cluster.local/apps/spark" \
  --files src/main/resources/log4j.properties \
  --jars target/gemini-deps.jar,target/gemini-uber.jar
```

```scala
val files = repos.getHEAD
  .getCommits
  .getTreeEntries
  .getBlobs
  .filter('is_binary === false)

files
 .sort("repository_id")
 .coalesce(1000)
 .write.parquet("hdfs://hdfs-namenode.default.svc.cluster.local/pga2/parquet/files")
```
</details>

Caching all files on-disk in Parquet fails though, \w 
```
Job aborted due to stage failure: Total size of serialized results of XX tasks (1025 MB) is bigger than spark.driver.maxResultSize (1024.0 MB)
```

This happens due to the fact that [DF API for a String column](https://github.com/apache/spark/blob/833c8d41527aa7f083e52cb6ae6cb1042987f064/sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala#L273) keeps in-memory the longest string, which is a full file content and is more then 1Gb.

**Comment by bzz** [comment]
On-disk Parquet cache was failing due to the number of tasks ~40k beeing to high for our clusted configuration, was fixed by reducing the number and can proceed over full PGA (~50h) 🎉  but is failing at the end now 😖  with 


```
Job aborted due to stage failure: Task 49 in stage 4.1 failed 4 times, most recent failure:
Lost task 49.3 in stage 4.1 (TID 27912, 10.2.56.116, executor 1):
java.io.FileNotFoundException: /spark-temp-data/spark-fb7fbf1e-033e-4122-9464-16acdc52fe34/executor-801bb555-6e93-4c4e-b3f8-46bc33ca9639/blockmgr-309ecee9-419e-45ab-a595-03dc3157b641/26/temp_shuffle_2eddf3bb-b5c0-488c-b397-47e4e4921a32 
(No such file or directory)
```

<details>

Simple example to reproduce

```scala
val path = "hdfs://hdfs-namenode.default.svc.cluster.local/pga2/siva/latest/"
val engine = Engine(spark, path, "siva")
val repos = engine.getRepositories

val files = repos.getHEAD
  .getCommits
  .getTreeEntries
  .getBlobs
  .filter('is_binary === false)

files
 .coalesce(1000)
 .sort("repository_id")
 .write.parquet("hdfs://hdfs-namenode.default.svc.cluster.local/pga2/parquet/files")
```
</details>
